diff --git a/src/macappkit.m b/src/macappkit.m
index 1d9b6938db..8ee264ce4e 100644
--- a/src/macappkit.m
+++ b/src/macappkit.m
@@ -1924,6 +1924,18 @@ - (BOOL)doesHoldQuit
 
 @implementation EmacsWindow
 
++ (NSButton *)standardWindowButton:(NSWindowButton)b forStyleMask:(NSWindowStyleMask)styleMask
+{
+
+  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+  NSString *hideDocumentIcon = [userDefaults stringForKey: @"HideDocumentIcon"];
+
+  if([hideDocumentIcon isEqualToString: @"YES"] && (b == NSWindowDocumentIconButton || b == NSWindowDocumentVersionsButton)) {
+    return nil;
+  }
+  return [NSWindow standardWindowButton:b forStyleMask:styleMask];
+}
+
 - (instancetype)initWithContentRect:(NSRect)contentRect
 			  styleMask:(NSWindowStyleMask)windowStyle
 			    backing:(NSBackingStoreType)bufferingType
@@ -2484,6 +2496,20 @@ - (void)setupWindow
   window.contentView = visualEffectView;
   MRC_RELEASE (visualEffectView);
   FRAME_BACKGROUND_ALPHA_ENABLED_P (f) = true;
+
+  NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+  NSString *transparentTitleBar = [userDefaults stringForKey: @"TransparentTitleBar"];
+
+  if ([transparentTitleBar isEqualToString: @"DARK"]) {
+    window.titlebarAppearsTransparent = true;
+    window.appearance = [NSAppearance appearanceNamed: NSAppearanceNameVibrantDark];
+  }
+
+  if ([transparentTitleBar isEqualToString: @"LIGHT"]) {
+    window.titlebarAppearsTransparent = true;
+    window.appearance = [NSAppearance appearanceNamed: NSAppearanceNameVibrantLight];
+  }
+
   if (FRAME_MAC_DOUBLE_BUFFERED_P (f))
     {
       FRAME_SYNTHETIC_BOLD_WORKAROUND_DISABLED_P (f) = true;
@@ -6297,7 +6323,7 @@ + (void)initialize
       NSUserDefaults *defaults = NSUserDefaults.standardUserDefaults;
 
       if ([defaults objectForKey:@"ApplePressAndHoldEnabled"] == nil)
-	[defaults registerDefaults:@{@"ApplePressAndHoldEnabled" : @"NO"}];
+	[defaults registerDefaults:@{@"ApplePressAndHoldEnabled" : @"NO", @"TransparentTitleBar": @"NO", @"HideDocumentIcon": @"NO"}];
     }
 }
 
